### PR TITLE
EIPs 155, 160, 161

### PR DESCRIPTION
**Pull request by gavofyork**
Components:
- [x] [EIP-155](https://github.com/ethereum/EIPs/issues/155)
- [x] [EIP-160](https://github.com/ethereum/EIPs/issues/160)
- [x] [EIP-161abc](https://github.com/ethereum/EIPs/issues/161)
- [x] [EIP-161d](https://github.com/ethereum/EIPs/issues/161)
- [x] EIP-155 tests
- [x] Integrate new JSON consensus tests
- [x] Finalise block number for transition

### CODE REVIEW

File: ethcore/src/types/transaction.rs [hunk 1]
```diff
@@ -72,8 +72,8 @@ pub struct Transaction {
 
 impl Transaction {
 	/// Append object with a without signature into RLP stream
-	pub fn rlp_append_unsigned_transaction(&self, s: &mut RlpStream) {
-		s.begin_list(6);
+	pub fn rlp_append_unsigned_transaction(&self, s: &mut RlpStream, network_id: Option<u8>) {
```
Comment (arkpar) [inline ethcore/src/types/transaction.rs]: Should be made private so that it is not misused for RLP serialization

**Comment by rphmeier** [comment]
All these spec file keys are gonna be a bit hellish for users to migrate their private chains.

File: ethcore/src/state/substate.rs [hunk 2]
```diff
@@ -45,10 +50,20 @@ impl Substate {
 	/// Merge secondary substate `s` into self, accruing each element correspondingly.
 	pub fn accrue(&mut self, s: Substate) {
 		self.suicides.extend(s.suicides.into_iter());
+		self.garbage.extend(s.garbage.into_iter());
```
Comment (rphmeier) [inline ethcore/src/state/substate.rs]: is it guaranteed that the union of two substates won't contain the same address in two sets?

Comment (rphmeier) [inline ethcore/src/state/substate.rs]: (in expected usage)

File: ethcore/src/types/transaction.rs [hunk 3]
```diff
@@ -135,26 +140,26 @@ impl From<ethjson::transaction::Transaction> for SignedTransaction {
...
 		let mut stream = RlpStream::new();
-		self.rlp_append_unsigned_transaction(&mut stream);
+		self.rlp_append_unsigned_transaction(&mut stream, network_id);
 		stream.out().sha3()
 	}
 
 	/// Signs the transaction as coming from `sender`.
-	pub fn sign(self, secret: &Secret) -> SignedTransaction {
-		let sig = ::ethkey::sign(secret, &self.hash())
+	pub fn sign(self, secret: &Secret, network_id: Option<u8>) -> SignedTransaction {
```
Comment (gavofyork) [inline ethcore/src/types/transaction.rs]: standard would be a misnomer - could be `sign_without_network_id`, but then i think `sign(..., None)` says mostly the same without the verbosity and additional `fn`.

File: ethcore/src/types/transaction.rs [hunk 4]
```diff
@@ -198,13 +203,20 @@ impl Transaction {
 		Self::gas_required_for(match self.action{Action::Create=>true, Action::Call(_)=>false}, &self.data, schedule)
 	}
 }
-
+/*
+enum NetworkRequirement {
+	Global,
+	Local(u8),
+	Either(u8),
+}
```
Comment (gavofyork) [inline ethcore/src/types/transaction.rs]: needs removing.

**Comment by gavofyork** [comment]
yeah - it's not our fault, though. @keorn has a script to help with this.

**Pull request merged**